File: Problem/By section/Dynamic Programing/Problem 343/problem_343.java
class Solution {
    // dp[i] = the maximum product of i
    // update rule
    // dp[i] = argmax_j (dp[i - j] * dp[j])
    // Note here: if j <= 4, dp[j] <= j
    // Also we can cut half of the searching space because
    // eg. dp[2] * dp[4] = dp[4] * dp[2]
    private int[] dp;
    
    // Time: O(n^2), Space: O(n)
    public int integerBreak(int n) {
        if(n <= 1) return 0;
        
        dp = new int[n + 1];
        dp[0] = 0;
        dp[1] = 1;
        dp[2] = 1;
        
        for(int i = 3; i <= n; i++){
            int limit = (int)(i / 2);
            
            for(int j = 1; j <= limit; j++){
                int temp_1 = ((i - j) > 4) ? dp[i - j] : (i - j);
                int temp_2 = (j > 4) ? dp[j] : j;
                dp[i] = Math.max(dp[i], temp_1 * temp_2);
            }
        }
        return dp[n];
    }
}

File: Problem/By section/Dynamic Programing/Problem 343/problem_343.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // dp[i] = the maximum product of i
    // update rule
    // dp[i] = argmax_j (dp[i - j] * dp[j])
    // Note here: if j <= 4, dp[j] <= j
    // Also we can cut half of the searching space because
    // eg. dp[2] * dp[4] = dp[4] * dp[2]
    vector<int> dp;
    
    // Time: O(n^2), Space: O(n)
    int integerBreak(int n) {
        dp = vector<int>(n, INT_MIN);
        dp[0] = 1;
        
        for(int i = 2; i <= n; i++) {
            for(int j = 1; j <= (int)(i/2); j++) {
                int one_part = max(dp[j - 1], j);
                int another_part = max(dp[i - j - 1], i - j);
                
                dp[i - 1] = max(dp[i - 1], one_part * another_part);
            }
        }
        
        return dp[n - 1];
    }
};

int main(int argc, char** argv) {
    Solution s;
    
    return 0;
}